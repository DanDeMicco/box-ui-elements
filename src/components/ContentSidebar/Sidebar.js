/**
 * @flow
 * @file Preview sidebar component
 * @author Box
 */

import * as React from 'react';
import DetailsSidebar from './DetailsSidebar';
import SkillsSidebar from './SkillsSidebar';
import ActivitySidebar from './ActivitySidebar';
import SidebarNav from './SidebarNav';
import { SIDEBAR_VIEW_SKILLS, SIDEBAR_VIEW_ACTIVITY, SIDEBAR_VIEW_DETAILS } from '../../constants';
import type { DetailsSidebarProps } from './DetailsSidebar';
import type { ActivitySidebarProps } from './ActivitySidebar';
import './Sidebar.scss';

type Props = {
    view?: SidebarView,
    currentUser?: User,
    file: BoxItem,
    getPreviewer: Function,
    activitySidebarProps: ActivitySidebarProps,
    detailsSidebarProps: DetailsSidebarProps,
    hasSkills: boolean,
    hasMetadata: boolean,
    hasActivityFeed: boolean,
    hasSkills: boolean,
    hasDetails: boolean,
    onSkillChange: Function,
    getApproverWithQuery?: Function,
    getMentionWithQuery?: Function,
    translations?: Translations,
    approverSelectorContacts?: SelectorItems,
    mentionSelectorContacts?: SelectorItems,
    activityFeedError?: Errors,
    currentUserError?: Errors,
    getAvatarUrl: (string) => Promise<?string>,
    onToggle: Function,
    onVersionHistoryClick?: Function,
    feedItems?: FeedItems
};

const Sidebar = ({
    view,
    currentUser,
    file,
    getPreviewer,
    hasMetadata,
    hasActivityFeed,
    hasSkills,
    hasDetails,
    activitySidebarProps,
    detailsSidebarProps,
    onSkillChange,
    getApproverWithQuery,
    getMentionWithQuery,
    activityFeedError,
    approverSelectorContacts,
    mentionSelectorContacts,
    getAvatarUrl,
    onToggle,
    onVersionHistoryClick,
    feedItems
}: Props) => (
    <React.Fragment>
        <SidebarNav
            onToggle={onToggle}
            selectedView={view}
            hasSkills={hasSkills}
            hasMetadata={hasMetadata}
            hasActivityFeed={hasActivityFeed}
            hasDetails={hasDetails}
        />
        {view === SIDEBAR_VIEW_DETAILS &&
            hasDetails && (
                <DetailsSidebar
                    file={file}
                    onVersionHistoryClick={onVersionHistoryClick}
                    {...detailsSidebarProps}
                />
            )}
        {view === SIDEBAR_VIEW_SKILLS &&
            hasSkills && <SkillsSidebar file={file} getPreviewer={getPreviewer} onSkillChange={onSkillChange} />}
        {view === SIDEBAR_VIEW_ACTIVITY &&
            hasActivityFeed && (
                <ActivitySidebar
                    currentUser={currentUser}
                    file={file}
                    activityFeedError={activityFeedError}
                    approverSelectorContacts={approverSelectorContacts}
                    mentionSelectorContacts={mentionSelectorContacts}
                    getApproverWithQuery={getApproverWithQuery}
                    getMentionWithQuery={getMentionWithQuery}
                    getAvatarUrl={getAvatarUrl}
                    onVersionHistoryClick={onVersionHistoryClick}
                    feedItems={feedItems}
                    {...activitySidebarProps}
                />
            )}
    </React.Fragment>
);

export default Sidebar;
