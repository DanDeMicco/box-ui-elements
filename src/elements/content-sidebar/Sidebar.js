/**
 * @flow
 * @file Preview sidebar component
 * @author Box
 */

import * as React from 'react';
import SidebarUtils from './SidebarUtils';
import {
    SIDEBAR_VIEW_SKILLS,
    SIDEBAR_VIEW_ACTIVITY,
    SIDEBAR_VIEW_DETAILS,
    SIDEBAR_VIEW_METADATA,
    ORIGIN_DETAILS_SIDEBAR,
    ORIGIN_ACTIVITY_SIDEBAR,
    ORIGIN_SKILLS_SIDEBAR,
    ORIGIN_METADATA_SIDEBAR,
} from '../../constants';
import type { DetailsSidebarProps } from './DetailsSidebar';
import type { ActivitySidebarProps } from './ActivitySidebar';
import type { MetadataSidebarProps } from './MetadataSidebar';
import './Sidebar.scss';

type Props = {
    activitySidebarProps: ActivitySidebarProps,
    currentUser?: User,
    detailsSidebarProps: DetailsSidebarProps,
    file: BoxItem,
    fileId: string,
    getPreview: Function,
    getViewer: Function,
    metadataSidebarProps: MetadataSidebarProps,
    onVersionHistoryClick?: Function,
    selectedView?: SidebarView,
};

// TODO: place into code splitting logic
const BASE_EVENT_NAME = '_JS_LOADING';
const MARK_NAME_JS_LOADING_DETAILS = `${ORIGIN_DETAILS_SIDEBAR}${BASE_EVENT_NAME}`;
const MARK_NAME_JS_LOADING_ACTIVITY = `${ORIGIN_ACTIVITY_SIDEBAR}${BASE_EVENT_NAME}`;
const MARK_NAME_JS_LOADING_SKILLS = `${ORIGIN_SKILLS_SIDEBAR}${BASE_EVENT_NAME}`;
const MARK_NAME_JS_LOADING_METADATA = `${ORIGIN_METADATA_SIDEBAR}${BASE_EVENT_NAME}`;

const LoadableDetailsSidebar = SidebarUtils.getAsyncSidebarContent(SIDEBAR_VIEW_DETAILS, MARK_NAME_JS_LOADING_DETAILS);
const LoadableActivitySidebar = SidebarUtils.getAsyncSidebarContent(
    SIDEBAR_VIEW_ACTIVITY,
    MARK_NAME_JS_LOADING_ACTIVITY,
);
const LoadableSkillsSidebar = SidebarUtils.getAsyncSidebarContent(SIDEBAR_VIEW_SKILLS, MARK_NAME_JS_LOADING_SKILLS);
const LoadableMetadataSidebar = SidebarUtils.getAsyncSidebarContent(
    SIDEBAR_VIEW_METADATA,
    MARK_NAME_JS_LOADING_METADATA,
);

const Sidebar = ({
    activitySidebarProps,
    currentUser,
    detailsSidebarProps,
    file,
    fileId,
    getPreview,
    getViewer,
    metadataSidebarProps,
    onVersionHistoryClick,
    selectedView,
}: Props) => (
    <React.Fragment>
        {selectedView === SIDEBAR_VIEW_DETAILS && (
            <LoadableDetailsSidebar
                key={fileId}
                fileId={fileId}
                onVersionHistoryClick={onVersionHistoryClick}
                {...detailsSidebarProps}
                startMarkName={MARK_NAME_JS_LOADING_DETAILS}
            />
        )}
        {selectedView === SIDEBAR_VIEW_SKILLS && (
            <LoadableSkillsSidebar
                key={file.id}
                file={file}
                getPreview={getPreview}
                getViewer={getViewer}
                startMarkName={MARK_NAME_JS_LOADING_SKILLS}
            />
        )}
        {selectedView === SIDEBAR_VIEW_ACTIVITY && (
            <LoadableActivitySidebar
                currentUser={currentUser}
                file={file}
                onVersionHistoryClick={onVersionHistoryClick}
                {...activitySidebarProps}
                startMarkName={MARK_NAME_JS_LOADING_ACTIVITY}
            />
        )}
        {selectedView === SIDEBAR_VIEW_METADATA && (
            <LoadableMetadataSidebar
                fileId={fileId}
                {...metadataSidebarProps}
                startMarkName={MARK_NAME_JS_LOADING_METADATA}
            />
        )}
    </React.Fragment>
);

export default Sidebar;
