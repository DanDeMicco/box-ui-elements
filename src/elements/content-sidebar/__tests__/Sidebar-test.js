import React from 'react';
import { shallow } from 'enzyme';
import Sidebar from '../Sidebar';

jest.mock('../../common/async-load', () => () => 'LoadableComponent');

describe('elements/content-sidebar/Sidebar', () => {
    const file = { id: 'id' };
    const getWrapper = props => shallow(<Sidebar file={file} {...props} />);

    test('should render no sidebar', () => {
        const wrapper = getWrapper();
        expect(wrapper).toMatchSnapshot();
    });

    test('should render skills sidebar', () => {
        const wrapper = getWrapper({ hasSkills: true, selectedView: 'skills' });
        expect(wrapper).toMatchSnapshot();
    });

    test('should render activity sidebar', () => {
        const wrapper = getWrapper({ hasActivityFeed: true, selectedView: 'activity' });
        expect(wrapper).toMatchSnapshot();
    });

    test('should render details sidebar', () => {
        const wrapper = getWrapper({ hasDetails: true, selectedView: 'details' });
        expect(wrapper).toMatchSnapshot();
    });

    test('should render metadata sidebar', () => {
        const wrapper = getWrapper({ hasMetadata: true, selectedView: 'metadata' });
        expect(wrapper).toMatchSnapshot();
    });
});
