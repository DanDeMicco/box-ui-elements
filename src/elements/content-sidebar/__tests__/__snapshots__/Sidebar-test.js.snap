// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/content-sidebar/Sidebar should render activity sidebar 1`] = `
<Fragment>
  <LoadableComponent
    file={
      Object {
        "id": "id",
      }
    }
    startMarkName="activity_sidebar_JS_LOADING"
  />
</Fragment>
`;

exports[`elements/content-sidebar/Sidebar should render details sidebar 1`] = `
<Fragment>
  <LoadableComponent
    startMarkName="details_sidebar_JS_LOADING"
  />
</Fragment>
`;

exports[`elements/content-sidebar/Sidebar should render metadata sidebar 1`] = `
<Fragment>
  <LoadableComponent
    startMarkName="metadata_sidebar_JS_LOADING"
  />
</Fragment>
`;

exports[`elements/content-sidebar/Sidebar should render no sidebar 1`] = `<Fragment />`;

exports[`elements/content-sidebar/Sidebar should render skills sidebar 1`] = `
<Fragment>
  <LoadableComponent
    file={
      Object {
        "id": "id",
      }
    }
    key="id"
    startMarkName="skills_sidebar_JS_LOADING"
  />
</Fragment>
`;
